core/lib/spree/bus.rb

```diff
@@ -0,0 +1,20 @@
+require 'omnes'
+
+module Spree
+  # Global [Omnes](https://github.com/nebulab/omnes) bus.
+  #
+  # This is used for internal events, while host applications are also able to
+  # use it.
+  #
+  # It has some modifications to support internal usage of the legacy event
+  # system {see Spree::AppConfiguration#use_legacy_events}.
+  Bus = Omnes::Bus.new.tap do |bus|
+    bus.define_singleton_method(:publish) do |*args, **kwargs, &block|
+      if Spree::Config.use_legacy_events
+        Spree::Event.fire(*args, **kwargs, &block)
+      else
+        super(*args, **kwargs, caller_location: caller_locations(1)[0], &block)
+      end
+    end
+  end
```

Comment:
Not a strong opinion, but I know `tap` can be a stumbling block for people not familiar with it. I like `tap`, but only use it when it actually makes the code cleaner or avoids polluting a scope with too many local variables. The use here doesn't really seem to add much to me, since we could write it like this instead and I think it's just as good, but doesn't require you to be familiar with `tap`:

```suggestion
  Bus = Omnes::Bus.new
  Bus.define_singleton_method(:publish) do |*args, **kwargs, &block|
    if Spree::Config.use_legacy_events
      Spree::Event.fire(*args, **kwargs, &block)
    else
      super(*args, **kwargs, caller_location: caller_locations(1)[0], &block)
    end
  end
```

Reply:
Fair enough. Changed. Thanks, @jarednorman.